whisper.cpp

```diff
@@ -2347,17 +2405,17 @@ static struct ggml_cgraph * whisper_build_graph_decoder(
 
     struct ggml_tensor * embd = ggml_new_tensor_1d(ctx0, GGML_TYPE_I32, n_tokens);
     ggml_set_name(embd, "embd");
-    ggml_set_input(embd);
+    //ggml_set_input(embd);
 
     struct ggml_tensor * position = ggml_new_tensor_1d(ctx0, GGML_TYPE_I32, n_tokens);
     ggml_set_name(position, "position");
-    ggml_set_input(position);
+    //ggml_set_input(position);
 
     const float KQscale = pow(float(n_state_head), -0.25);
 
     struct ggml_tensor * KQ_mask = ggml_new_tensor_3d(ctx0, GGML_TYPE_F32, n_kv, GGML_PAD(n_tokens, GGML_KQ_MASK_PAD), 1);
     ggml_set_name(KQ_mask, "KQ_mask");
-    ggml_set_input(KQ_mask);
+    //ggml_set_input(KQ_mask);
```

Comment:
When running with the Metal backend, having these `ggml_set_input()` calls in the decoder graph causes the assigned backends to be NULL (nodes 0, 1, 27):

```
## SPLIT #0: Metal # 0 inputs: 
node #  0 (  GET_ROWS):               node_0 (   2K) [Metal   1.wgt0]:               leaf_1 (  50M) [Metal    1.dst]         Metal#embd#0 (   1K) [ NULL    4.cpy]
node #  1 (  GET_ROWS):               node_1 (   2K) [Metal   1.wgt0]:               leaf_3 ( 896K) [Metal    1.dst]     Metal#position#0 (   1K) [ NULL    4.cpy]
node #  2 (       ADD):               node_2 (   2K) [Metal    2.sup]:               node_0 (   2K) [Metal   1.wgt0]               node_1 (   2K) [Metal   1.wgt0]
node #  3 (      NORM):               node_3 (   2K) [Metal    2.sup]:               node_2 (   2K) [Metal    2.sup]
node #  4 (       MUL):               node_4 (   2K) [Metal   1.wgt1]:               node_3 (   2K) [Metal    2.sup]               leaf_5 (   2K) [Metal    1.dst]
node #  5 (       ADD):               node_5 (   2K) [Metal   1.wgt1]:               node_4 (   2K) [Metal   1.wgt1]               leaf_6 (   2K) [Metal    1.dst]
node #  6 (   MUL_MAT):               node_6 (   2K) [Metal   1.wgt0]:               leaf_0 ( 512K) [Metal    1.dst]               node_5 (   2K) [Metal   1.wgt1]
node #  7 (     SCALE):               node_7 (   2K) [Metal    2.sup]:               node_6 (   2K) [Metal   1.wgt0]
node #  9 (       CPY): leaf_7 (view) (copy  (   1K) [Metal   1.vsrc]:               node_7 (   2K) [Metal    2.sup]        leaf_7 (view) (   1K) [Metal   1.vsrc]
node # 10 (   MUL_MAT):              node_10 (   2K) [Metal   1.wgt0]:               leaf_8 ( 512K) [Metal    1.dst]               node_5 (   2K) [Metal   1.wgt1]
node # 11 (       ADD):              node_11 (   2K) [Metal   1.wgt1]:              node_10 (   2K) [Metal   1.wgt0]               leaf_9 (   2K) [Metal    1.dst]
node # 15 (       CPY): leaf_10 (view) (copy (   1M) [Metal   1.vsrc]:  (reshaped) (transpo (   2K) [Metal   4.vsrc]       leaf_10 (view) (   1M) [Metal   1.vsrc]
node # 21 (   MUL_MAT):              node_21 (   2K) [Metal   1.wgt0]:              leaf_19 ( 512K) [Metal    1.dst]               node_5 (   2K) [Metal   1.wgt1]
node # 22 (       ADD):              node_22 (   2K) [Metal   1.wgt1]:              node_21 (   2K) [Metal   1.wgt0]              leaf_20 (   2K) [Metal    1.dst]
node # 23 (     SCALE):              node_23 (   2K) [Metal    2.sup]:              node_22 (   2K) [Metal   1.wgt1]
node # 26 (   MUL_MAT):              node_26 (   0K) [Metal    2.sup]:        leaf_7 (view) (   1K) [Metal   1.vsrc]  (reshaped) (permute (   2K) [Metal   4.vsrc]
node # 27 (  SOFT_MAX):              node_27 (   0K) [Metal    2.sup]:              node_26 (   0K) [Metal    2.sup]      Metal#KQ_mask#0 (   2M) [ NULL    4.cpy]
node # 28 (   MUL_MAT):              node_28 (   2K) [Metal    2.sup]:       leaf_10 (view) (   1M) [Metal   1.vsrc]              node_27 (   0K) [Metal    2.sup]
```

Is this expected? Without Metal it works as expected. Removing the calls also seems to work, but seems incorrect since the `KQ_mask` might get overwritten?

Reply:
I am not sure what's happening. When I uncomment this line, I get `ggml_metal_get_buffer: error: tensor '��M9' buffer is nil`, and the garbled name suggests memory corruption.

Reply:
It's the same here. I thought the source of the issue is the `NULL` backend assigned for these nodes

Reply:
The `NULL` is expected here, it's because split inputs do not have a backend assignment in the hashtable. This is just so that the size of the hash table remains predictable. I think the tensors structs are being overwritten, but I am not sure where.